packages/web-components/fast-components/src/tabs/tabs.ts

```diff
@@ -0,0 +1,301 @@
+            this.setTabs();
+            this.setTabPanels();
+            this.handleActiveIndicatorPosition();
+        }
+    }
+
+    public start: HTMLSlotElement;
+    public startContainer: HTMLDivElement;
+    public handleStartContentChange(): void {
+        this.start.assignedNodes().length > 0
+            ? this.startContainer.classList.add("start")
+            : this.startContainer.classList.remove("start");
+    }
+
+    public end: HTMLSlotElement;
+    public endContainer: HTMLDivElement;
+    public handleEndContentChange(): void {
+        this.end.assignedNodes().length > 0
+            ? this.endContainer.classList.add("end")
+            : this.endContainer.classList.remove("end");
+    }
+
+    @attr({ mode: "boolean" })
+    public activeindicator = true;
+
+    @observable
+    public activeIndicatorRef: HTMLElement;
+    public activetab: HTMLElement;
+
+    private prevActiveTabIndex: number = 0;
+    private activeTabIndex: number = 0;
+    private ticking: boolean = false;
+    private tabIds: Array<string | null>;
+    private tabpanelIds: Array<string | null>;
+    private connected: boolean = false;
+
+    private change = (): void => {
+        this.$emit("change", this.activetab);
+    };
+
+    private getActiveIndex(): number {
+        if (this.activeid !== undefined) {
+            return this.tabIds.indexOf(this.activeid) === -1
+                ? 0
+                : this.tabIds.indexOf(this.activeid);
+        } else {
+            return 0;
+        }
+    }
+
+    private setTabs = (): void => {
+        this.tabIds = this.getTabIds();
+        this.tabpanelIds = this.getTabPanelIds();
+        this.activeTabIndex = this.getActiveIndex();
+        this.tabs.forEach((tab: HTMLElement, index: number) => {
+            if (tab.slot === "tab") {
+                tab.setAttribute(
+                    "id",
+                    this.tabIds[index] === undefined || this.tabIds[index] === null
+                        ? `tab-${index + 1}`
+                        : (this.tabIds[index] as string)
+                );
+                tab.setAttribute(
+                    "aria-selected",
+                    this.activeTabIndex === index ? "true" : "false"
+                );
+                tab.setAttribute(
+                    "aria-controls",
+                    this.tabpanelIds[index] === undefined ||
+                        this.tabpanelIds[index] === null
+                        ? `panel-${index + 1}`
+                        : (this.tabpanelIds[index] as string)
+                );
+                tab.setAttribute(
+                    "style",
+                    this.isHorizontal()
+                        ? `grid-column: ${index + 1};`
+                        : `grid-row: ${index + 1};`
+                );
+                tab.addEventListener("click", this.handleTabClick);
+                tab.addEventListener("keydown", this.handleTabKeyDown);
```

Comment:
Should we do something to prevent multiple even listener attach if this method is called more than once?

Reply:
Another way to handle this might be to have each tab handle its own click and keydown event, and then simply emit a custom event that the tab component can handle to control what happens. If we did that, you can add all event handlers in the constructor. This won't work if you want to handle any type of element being a tab though. We'd have to keep it as is for that scenario. In which case, you'll want to tag the elements to make sure you don't add duplicate handlers as mentioned above.

Reply:
Rob and I chatted about this offline. If you give it the same exact instance for the same event, it won't get registered twice. So we are good.